Make end-to-end Tests for Editable Textfield

schen2166 opened the pull request:
Enter:
Makes sure that any edits save after a Enter key press
Tests the title saves current title if Only Enter is Pressed

ESC:
Tests if ESC key press will cancel any edits made. reverts to original title before edit
Tests text field will remains same if only ESC is pressed.

Also changed to afterEach in each test suite to make sure each test runs with the same default parameters.

Note: When reviewed, may want to combine this cypress test with OSCAL_Edit_Tests.cy.js

Closes the Issue https://github.com/EasyDynamics/oscal-editor-deployment/issues/97

tuckerzp commented:
A few things just to keep on track:
1. Are your local `Edit` tests still failing? They seem to be passing when run through the workflow.
2. I personally think that keeping the `TextField` tests inside the confines of `LoadTests` & `EditTests` makes the most sense. 
3. I am not sure if linting code is the best thing to be doing for this specific pr. I am not against it completely, but it makes the `diff` a bit tricky to understand what is actually being changed.

Commit:
Add cypress tests for textfield

resetted test defaults with afterEach

Test ESC and ENTER e2e Cypress Tests

Creates cypress tests to test functionality of Enter key
Edits the textfield and when on enter keypress, the
value of the textfield becomes the current value of the textfield

ESC checks if the value of the textbox remains the same.
Essentially the ESC key cancels the editing of the textbox.

Update tests to to set test state before each test

Before() and after() functions refreshes the state for each test
so that each of the tests can properly test their functions.
The tests don't reset their initial state after a test case.

update test based on the edit_tests.cy.js

reformatted files using prettier

made enter a trigger event instead of type

deleted clear (logic error)

Add AfterEach to reset unit test defaults

merged Testfield Tests with Edit tests

Delete textfield test merged w/ Edit Tests

End-to-End Test for Editbable Textfield

Edit_Tests:
Added the tests for the Enter and ESC keys
-Enter Key
        Tests for the textfield to save changes for both
        when the field is edited or left unchanged

-ESC
        Tests that the textfield remains the same on ESC keypress

Note: SSP_TITLE_ORIG constant was changed to 'Another SSP' to
comply with the test. Other assertions were changed accordingly
to make sure properties are still being tested

Load_Tests:
Changed value of constant SSP_TITLE_ORIG (see note above)

OSCAL_Textfield Tests:
Changed to `AfterEach` to make sure the testing environment
is the same for each test in the test suite.

Update tests to make sure it works. Contents in this file have been
copied to `OSCAL_Edit_Tests`

This serves as a backup copy for the tests

Fix testing errors

schen2166 commented:
Created end-to-end tests for Enter and Esc buttons
Enter:
Makes sure that any edits save after a Enter key press
Tests the title saves current title if Only Enter is Pressed

ESC:
Tests if ESC key press will cancel any edits made. reverts to original title before edit
Tests text field will remains same if only ESC is pressed.

Also changed to afterEach in each test suite to make sure each test runs with the same default parameters.
Closes the Issue #97

tuckerzp commented:
Just to make reviewing this a bit easier, what are the titles of the tests that are actually being added?

schen2166 commented:
> Just to make reviewing this a bit easier, what are the titles of the tests that are actually being added?

There are a total of 2 test suites each with 2 tests
Suite 1 (titled Test Editing with Enter key)
Test 1: Successfully Edits Title with Enter
Test 2: Successfully keeps title on Enter key

Suite 2: (Titled Test Cancel Edit with ESC)
Test 1: Successfully cancels any changes on ESC keypress
Test 2 ESC key cancels textfield edit action

These lines of code are found from lines 139-199 on OSCAL_Edit_Test.cy.js

tuckerzp reviewed (dismissed):
Looks good to me! Thanks for your work on this @schen2166

kylelaker commented on end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Textfield_Tests.cy.js:
```diff
@@ -41,24 +38,34 @@ describe("Cancel on ESC Keypress", () => {
       .clear()
-      .type("{esc}")
-      .should("have.value", SSP_TITLE_ORIG);
+      .type("{esc}");
+      cy.get(`[aria-label="edit-system-security-plan-metadata-title"]`).click();
+      cy.get(`[data-testid="textField-system-security-plan-metadata-title"]`)
+      .click()
+      .should("have.value", SSP_ALTERNATE);
```
Formatting is very weird here. The `cy.get` lines likely need to be un-indented? Same in a few other places.

laurelmay reviewed (commented):
These tests _really_ heavily rely on `aria-label` to select the components. Is there a better way for us to be able to find these edit buttons? Or is the `aria-label` the best option we have right now.

Usually, we want to think about accessing elements in the way that an end user would interact with the page. Even users who would be using accessibility tools likely aren't purely navigating by `aria-label`.

I assume typically a user would find the control, the component, and the text they want to edit and click the nearest "edit" button to that. Does our current DOM structure allow for that?

If not, that's okay and we can stick with the `aria-label`s for now; but it is a larger fix that we need to be aware of needing to correct in the future.

tuckerzp commented:
> I believe currently, the only way to access editing on the buttons is to click on the edit button.

Would it be difficult to emulate clicking on the edit button then editing the nearest text field?

schen2166 commented:
Not sure what you mean by that but you can see visually below. If the window is small, the edit button would be overshadowed by other components which make in inaccessible. We could make the text box clickable to resolve this.
<img src ="https://user-images.githubusercontent.com/106985986/178065386-51de0c14-1ea6-47d3-a272-dff1315075b0.png" width = "700">
Image above is window in full screen with is fine. Below is the web app at a reduced size. The `Edit` button should be next to **"BASELINE"** but the window size knocks it down to the next line. Which is behind the **"Joint Task Force, Transformation Intiative"** box component which is not clickable.
<img src="https://user-images.githubusercontent.com/106985986/178065488-61895e75-d91a-472b-96b6-48b514f69866.png" width = "300">

Bronstrom committed:
Patch broken Docker Hub links

Bronstrom committed:
Format all-in-one/README.md

schen2166 committed:
Update tests to comply with new `command.js` functions

The function `navToTestSspRestMode()` was removed from
`command.js` so changed those function calls to navtoSspViewer

tuckerzp commented on end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:
```diff
@@ -43,78 +43,154 @@ describe('Test Editing System Security Plan Source', () => {
+  });
+});
+
+describe("Test Editing Title Using Enter key", () => {
```
```suggestion
describe("Editing Title Using Enter key", () => {
```

tuckerzp commented on end-to-end-tests/cypress/e2e/OSCAL/OSCAL_Edit_Tests.cy.js:
```diff
@@ -43,78 +43,154 @@ describe('Test Editing System Security Plan Source', () => {
+  });
+});
+
+describe("Cancel on ESC Keypress", () => {
```
The wording for these tests are a bit confusing when reading the output. 

`Cancel on ESC Keypress ESC key cancels any changes`
`Cancel on ESC Keypress Enter key changes textfield value`

Is there a way to rename them so they are a bit more descriptive of what is being tested? I think its mostly the `describe` that makes it confusing.

schen2166 replied:
Sure thing, I can rename the tests. I was unsure if you have already renamed them in your request to change the test names in your other pull request. I can scrap the names have at the moment and reword them and make them more clear what the tests are doing.

Are the `Enter` and `Esc` test suites in the other pull request which your edit the test titles?

tuckerzp replied:
If you look at #111, I have updated many of the tests made before your changes. As your `Editable Textfield` tests are not currently in `main`, I did not touch them

tuckerzp reviewed (changes requested):
I have a few comments on the naming of the tests. I think it is important to be able to understand the intention of the test without needing to look at code for the output logs.

schen2166 committed:
Change Text test name titles for clarity

schen2166 committed:
Merge remote-tracking branch 'origin' into feature/textfield-cypress-e2e-tests

schen2166 committed:
Change Function calls to `navToSspEditor()`

`navToSspViewer()` is no longer a function in `commands.js`
it has been renamed as `navToSspEditor()`

tuckerzp reviewed (approved):
Thanks for your work on this @schen2166! I am happy it will finally be merged in.

laurelmay reviewed (approved):
Thanks for these! Looks pretty good!
